server/rest/src/main/java/org/infinispan/rest/resources/ServerResource.java

```diff
@@ -332,26 +344,82 @@ private CompletionStage<RestResponse> threads(RestRequest restRequest) {
                if (t != null) {
-                  return responseBuilder.status(HttpResponseStatus.INTERNAL_SERVER_ERROR).build();
-               } else {
-                  return responseBuilder
-                        .contentType(MediaType.fromString("application/gzip"))
-                        .header("Content-Disposition",
-                              String.format("attachment; filename=\"%s-%s-%3$tY%3$tm%3$td%3$tH%3$tM%3$tS-report.tar.gz\"",
-                                    Version.getBrandName().toLowerCase().replaceAll("\\s", "-"),
-                                    invocationHelper.getRestCacheManager().getNodeName(),
-                                    Calendar.getInstance())
-                        )
-                        .entity(path.toFile()).build();
+                  return new NettyRestResponse.Builder().status(HttpResponseStatus.INTERNAL_SERVER_ERROR).build();
                }
+
+               return createReportResponse(path.toFile(), invocationHelper.getRestCacheManager().getNodeName());
             })
       );
    }
 
+   private RestResponse createReportResponse(Object report, String filename) {
+      return new NettyRestResponse.Builder()
+            .contentType(MediaType.fromString("application/gzip"))
+            .header("Content-Disposition",
+                  String.format("attachment; filename=\"%s-%s-%3$tY%3$tm%3$td%3$tH%3$tM%3$tS-report.tar.gz\"",
+                        Version.getBrandName().toLowerCase().replaceAll("\\s", "-"),
+                        filename,
+                        Calendar.getInstance())
+            )
+            .entity(report)
+            .build();
+   }
+
+   private CompletionStage<RestResponse> nodeReport(RestRequest restRequest) {
+      String targetName = restRequest.variables().get("nodeName");
+      EmbeddedCacheManager cacheManager = invocationHelper.getProtocolServer().getCacheManager();
+      NettyRestResponse.Builder responseBuilder = new NettyRestResponse.Builder();
+      List<Address> members = cacheManager.getMembers();
+
+      if (invocationHelper.getRestCacheManager().getNodeName().equals(targetName)) {
+         return report(restRequest);
+      }
+
+      if (members == null) {
+         return CompletableFuture.completedFuture(responseBuilder.status(NOT_FOUND).build());
+      }
+
+      ByRef<ByteBuffer> response = new ByRef<>(null);
+      return SecurityActions.getClusterExecutor(cacheManager)
+            .submitConsumer(ecm -> {
+               CacheManagerInfo cmi = ecm.getCacheManagerInfo();
+               if (!cmi.getNodeName().equals(targetName)) return null;
+
+               GlobalComponentRegistry gcr = SecurityActions.getGlobalComponentRegistry(ecm);
+               BasicComponentRegistry bcr = gcr.getComponent(BasicComponentRegistry.class);
+               ServerStateManager ssm = bcr.getComponent(ServerStateManager.class).wired();
+
+               CompletionStage<byte[]> report = ssm.managedServer().getServerReport()
```

Comment:
It would probably be a bit cleaner and most likely a tad more performant to instead do the await on the stage returned from `getServerReport` instead of chaining the method and thus performing the Files.readAllBytes on possibly a different thread.